core/src/main/java/hudson/model/Job.java

```diff
@@ -424,15 +427,24 @@ public synchronized void updateNextBuildNumber(int next) throws IOException {
     }
 
     /**
-     * Returns the configured build discarder for this job, or null if none.
+     * Returns the configured build discarder for this job, via {@link BuildDiscarderProperty}, or null if none.
      */
     public BuildDiscarder getBuildDiscarder() {
-        return logRotator;
+        BuildDiscarderProperty prop = getProperty(BuildDiscarderProperty.class);
```

Comment:
Why? Any synchronization is the responsibility of `getProperty` or the code it delegates to.